Telemetry changes:

msolivialuo opened the pull request:
1. Added the telemetry rule class and defined rules for each event property
2. Move all the collection logics out from the event classes and added to the aggregator class.
3. Renamed all the property names to have _PROPERTY_NAME and property values to have _PROPERTY_VALUE in the name. This way the names are self-descriptive and easy to tell which one is property name vs. value.
4. Changed “#define” to actual function. “#define” should not be used as it is not type safe and difficult to debug.

msftclas commented:
Hi __@msolivialuo__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Olivia Luo). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

RPangrle commented on ADAL/src/telemetry/ADTelemetryDefaultEvent.m:
```diff
@@ -132,13 +125,6 @@ - (NSString*)getStringFromDate:(NSDate*)date
-    SET_IF_NOT_NIL(eventToBeDispatched, AD_TELEMETRY_CORRELATION_ID, [_propertyMap objectForKey:AD_TELEMETRY_CORRELATION_ID]);
-}
-
 - (NSDictionary*)defaultParameters
```
Make this a class method

msolivialuo replied:
This is a class method. Am I missing anything??

RPangrle replied:
This got put on the wrong line somehow

RPangrle replied:
This comment is on `-defaultParameters` which is an instance method. It should be a class method.

RPangrle commented on ADAL/src/telemetry/ADTelemetryDefaultEvent.h:
```diff
@@ -30,14 +30,15 @@
     NSInteger _defaultPropertyCount;
 }
 
+- (NSDictionary*)getProperties;
```
Assuming this is an accessor, remove 'get' from the method name.

msolivialuo replied:
In the original code, this was a private method which returns a private dictionary. The change is to make the method public. I will keep it this way to avoid large changes.

msolivialuo replied:
Updated.

RPangrle commented on ADAL/src/telemetry/ADTelemetryDefaultEvent.h:
```diff
@@ -30,7 +30,7 @@
 }
 
-- (NSDictionary*)getProperties;
+@property (nonatomic, strong, readonly, getter=getProperties) NSDictionary *properties;
```
Getters should not have 'get' in them in objc

RPangrle commented on ADAL/src/telemetry/ADTelemetryEventStrings.h:
```diff
@@ -34,76 +34,77 @@ extern NSString *const AD_TELEMETRY_EVENT_TOKEN_CACHE_WRITE;
-extern NSString *const AD_TELEMETRY_EVENT_NAME;
-extern NSString *const AD_TELEMETRY_REQUEST_ID;
+
+extern NSString *const AD_TELEMETRY_PROPERTY_NAME_AUTHORITY_TYPE;
```
AD_TELEMETRY_PROPERTY_NAME => AD_TELEMETERY_KEY
AD_TELEMETRY_PROPERTY_VALUE => AD_TELEMETRY_VALUE

RPangrle commented on ADAL/src/utils/NSMutableDictionary+ADExtensions.m:
```diff
@@ -0,0 +1,37 @@
+
+@implementation NSMutableDictionary (ADAL)
+
+- (void)setObjectWithNilCheck:(id)dictObject forKey:(id<NSCopying>)dictKey
```
Change to `-adSetObjectIfNotNil:forKey:`

RPangrle commented on ADAL/src/telemetry/ADTelemetryDefaultEvent.m:
```diff
@@ -125,6 +125,19 @@ - (NSString*)getStringFromDate:(NSDate*)date
+    {
+        // Some telemetry collectors only allow alphabetic characters, underscores, and dots in the property names
+        [_propertyMap setObject:[propertyMapCopy objectForKey:propertyName]
+                         forKey:[propertyName stringByReplacingOccurrencesOfString:@"-" withString:@"_"]];
```
If we're going to be not allowing hyphens in property names/keys then we should be changing the key values themselves.

msolivialuo replied:
Checked with @jasoncoolmax before I made this change, these keys are not only used by telemetry, Changing these values will break existing apps.

msolivialuo commented:
Telemetry changes:
1. Added the telemetry rule class and defined rules for each event property
2. Move all the collection logics out from the event classes and added to the aggregator class.
3. Renamed all the property names to have _PROPERTY_NAME and property values to have _PROPERTY_VALUE in the name. This way the names are self-descriptive and easy to tell which one is property name vs. value.
4. Changed “#define” to actual function. “#define” should not be used as it is not type safe and difficult to debug.
